core/arch/interp.c

```diff
@@ -3683,6 +3695,12 @@ build_bb_ilist(dcontext_t *dcontext, build_bb_t *bb)
 # endif /* X86 */
 #endif /* WINDOWS */
 
+        if (dcontext->forged_exception_addr == bb->start_pc) {
```

Comment:
There's a do...while inside the while(true), at the top.

Given that we *know* this is the first instr maybe we can try to avoid paying for a check on every iteration of that inner loop -- maybe by making max_bb_instrs into a local that we set outside of both loops, and we set it to 1 for this single-step-bb case?